Handle drags when the view is nearly parallel to a drag plane

ericmehl opened the pull request:
This changes the behavior of `PlanarDrag` and `AngularDrag` to use a `LinearDrag` when the view is close to parallel to the plane being dragged. Previously, using Translate Tool and Rotate Tool would have no effect in such cases. 

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
Adapting the spot light tool to use the new behavior in the `AngularDrag` I realized I skated by on this change so far because the axes given to `AngularDrag` are always normalized. Not so with my current implementation of the spot light tool, so at a minimum I need to normalize a few vectors before doing cross and dot products on them.

danieldresser-ie commented on src/GafferUI/Handle.cpp:
```diff
@@ -556,18 +558,62 @@ Handle::AngularDrag::AngularDrag( bool processModifiers )
+		// Looking down the normal of the rotation plane, positive values rotate
+		// counterclockwise. Find the gadget-space vector pointing in the direction
+		// of positive rotation to use for the start direction of the linear drag.
+		V3f parallelAxis = useAxis0 ? axis1 : planeAxis0;
```
This naming feels confusing ... useAxis0 means it will use axis1, not axis0?

ericmehl replied:
Yes, "use" is too general here, is `int dragAxis = [0,1]` clearer? That's what I did in https://github.com/GafferHQ/gaffer/pull/5281/commits/b015786d886bd64a2e3299b445643ac3c05d4071

danieldresser-ie commented on src/GafferUI/Handle.cpp:
```diff
@@ -556,18 +558,62 @@ Handle::AngularDrag::AngularDrag( bool processModifiers )
+		// reverse the drag direction.
+		float sign = -( ( 0 < parallelDot ) - ( parallelDot < 0 ) );
+
+		V3f dragDirection = planeAxis0.cross( m_axis1 ).cross( parallelAxis );
```
I haven't fully wrapped my head around this, but if useAxis0 is true, this ends up setting drag direction to `-axis0.cross( axis1 ).cross( axis1 ).cross( axis1 )`. I feel like there's probably a way of expressing this without crossing with the same vector 3 times?

ericmehl replied:
Very nice, it turns out `planeAxis0.cross( m_axis1 )` is the same as `axis0`, so we can drop one or two crosses with that. Changed in https://github.com/GafferHQ/gaffer/pull/5281/commits/7696ac7a7149d6896d8a6c011448e00dae815329

danieldresser-ie commented on src/GafferUI/Handle.cpp:
```diff
@@ -71,6 +71,47 @@ using namespace IECore;
+	result.worldPlane = Plane3f(
+		result.worldOrigin,
+		result.worldOrigin + result.worldAxis0,
+		result.worldOrigin + result.worldAxis1
```
Gah, just lost a lengthy comment to webbrowser weirdness, I'll try to summarize over again:
I don't like using the 3 point version of the Plane3f constructor when we have directions to start with.
The Plane3f constructor just does:
```
normal = (point2 - point1) % (point3 - point1);
normal.normalize();
```
We can cancel out adding and subtracting worldOrigin and just do: `normal = ( worldAxis0 % worldAxis1 ).normalized()`, and in this case only the normal is used, so we don't need to store a Plane3f.

I would also say it would be better to compute the normal ourselves before passing it in to the temp Plane3f in updatePosition ... but this commit isn't touching that.

ericmehl replied:
Also a nice improvement, thanks for pointing that out. I removed the `Plane3f` use in the new code in https://github.com/GafferHQ/gaffer/pull/5281/commits/4cf983bd9790af0bd5f6f1b0b0af9c8c52580048.

In https://github.com/GafferHQ/gaffer/pull/5281/commits/3b696193e4074c399f803cda406288b35a10969f I strayed a bit out of scope of this PR and changed the `Plane3f` construction in `updatePosition()`. Not sure if it's a clear win as we drop a little calculation in exchange for another `V3f` stored in the class, but it's there for consideration.

danieldresser-ie reviewed (commented):
I was asked to take a look at the math, so I've done a little nitpicking. Probably none of it crucial though - I haven't actually played with it interactively to really understand the implications of how it works, but I think other folks are looking at that.

danieldresser-ie commented:
Those changes all look reasonable to me.

ericmehl commented:
I pushed a few new commits :
- https://github.com/GafferHQ/gaffer/pull/5281/commits/f859700b4d190ac7217a40d766a2dce4398bdb26 : Normalizes the vectors before taking the dot product. This is needed to be compatible with the comment in Handle.h `Axes are assumed to be orthogonal, but may have any length`
- https://github.com/GafferHQ/gaffer/pull/5281/commits/f6769bef3856cb4d96edf3f6ac3f7762e7ee3221 : Added `isLinearDrag()` which will be needed by the spotlight tool so it can modify its behavior according to the drag type.
- https://github.com/GafferHQ/gaffer/pull/5281/commits/8b8d009031d8903a837d820be08b6cbd25c9f081 : This changes the axis setup of `AngularDrag` to hopefully be more intuitive. I tested it out both for the `RotationTool` and on my in-progress `SpotLightTool` and it seems to do the job. And a small nice side effect of making a couple of the calculations simpler.

I believe that's all that is relevant to this PR for `AngleDrag` (when thinking about `SpotLightTool` at least) so it's ready for a full review.

ericmehl commented:
I rebased this onto `main` today and also updated the commit references in my comments above. It probably needs to be considered along with #5218 since there may be some architectural changes to that PR that belong here.

ericmehl commented:
I rebased onto the latest `main` and updated the comment resolution links to match. I think now that we have #5218 in a reviewable state, the review for this PR can move forward too.

danieldresser-ie commented:
Fixes all look good to me.

johnhaddon commented:
I've got a failure case for the rotation drags :

- Create a cube, and focus it.
- Look through the standard Front camera.
- Turn on the rotate tool, and drag the green handle for rotation about Y.
- Cube disappears, and nans are shown for the rotation in the NodeEditor.

ericmehl commented on src/GafferUI/Handle.cpp:
```diff
@@ -575,17 +575,17 @@ Handle::AngularDrag::AngularDrag( const Gadget *gadget, const Imath::V3f &origin
 		// If the parallel axis is pointing in the same direction as the view,
 		// reverse the drag direction.
-		float sign = -( ( 0 < parallelDot ) - ( parallelDot < 0 ) );
+		float parallelDot = dragAxis == 0 ? -axis90Dot : axis0Dot;
```
I don't quite understand the underlying math of why this negation is necessary unfortunately. I arrived at that result by charting out the different axis configurations and seeing which needed to be flipped, along with the `sign` below. I can certainly revise if there's a more elegant solution I'm missing.

johnhaddon replied:
@danieldresser-ie, could you take a look at this please? Since you reviewed originally I think you have a better chance of evaluating the new code...

danieldresser-ie replied:
I tried to figure out why the negative sign is necessary here, and didn't quite figure it out. But this approach seems to yield the same result, and I think is simpler?

```    if( abs( dragData.worldNormal.dot( dragData.worldLine.dir ) ) < g_planarToLinearDragThreshold )
    {
        // Do a LinearDrag in raster space instead of world space

        // We're going to find a direction in raster space where all movement along this axis will
        // be treated as rotation. We want a direction in the 2d plane that is perpendicular to the
        // normal, which we can get via cross product between the normal and the view direction.
        V3f worldRotationDirection = dragData.worldLine.dir.cross( dragData.worldNormal );

        // Convert to raster space
        const ViewportGadget *viewport = gadget->ancestor<ViewportGadget>();
        V2f rasterRotationDirection =
            ( viewport->worldToRasterSpace( worldRotationDirection + dragData.worldOrigin )
            - viewport->worldToRasterSpace( dragData.worldOrigin ) ).normalized()
            * V2f( 1.f, -1.f ) // Y is positive down in raster space, flip the Y direction
        ;

        m_drag = LinearDrag(
            gadget,
            rasterRotationDirection,
            dragBeginEvent,
            m_processModifiers
        );
        m_dragBeginRotation = std::get<LinearDrag>( m_drag ).startPosition();
    }
```

ericmehl replied:
That is much simpler and I confirmed that the rotation directions are right for all the directions I can think of to test.  Thanks Daniel!

Fixed in https://github.com/GafferHQ/gaffer/pull/5281/commits/154226b61ef5cd56754c803d42794b07fcec1131

ericmehl commented:
> I've got a failure case for the rotation drags

I hadn't quite gotten the conversion of the parameters right. I have a fix for that, and an invite for improvements, in https://github.com/GafferHQ/gaffer/pull/5281/commits/724761cf4fd9fca354afc65669a60758661eac31 now.

johnhaddon commented:
Fixes all seem to be working fine for me here. Could you squash down and merge please @ericmehl?

ericmehl commented:
Squashed and pushed! I'll just wait for CI to go through to be double safe
